packages/citizen-sdk/src/utils/auth.ts

```diff
@@ -1,6 +1,156 @@
 import { Account, createWalletClient, http } from "viem";
 import { Envs, FV_IDENTIFIER_MSG2 } from "../constants";
 
+// Cache for the Farcaster SDK to avoid repeated dynamic imports
+let _cachedSdk: typeof import('@farcaster/miniapp-sdk').sdk | null = null;
+
+/**
+ * Lazy-load and cache the Farcaster SDK
+ */
+export async function loadFarcasterSdk() {
+  if (!_cachedSdk) {
+    const { sdk } = await import('@farcaster/miniapp-sdk');
+    _cachedSdk = sdk;
+  }
+  return _cachedSdk;
+}
+
+/**
+ * Fallback detection using context check
+ */
+async function fallbackDetect(): Promise<boolean> {
+  try {
+    const sdk = await loadFarcasterSdk();
+    const context = await sdk.context;
+    return !!(context.location && context.location.type !== undefined);
+  } catch {
+    return false;
+  }
+}
+
+/**
+ * Detects if the SDK is running inside a Farcaster miniapp using the official SDK
+ */
+export async function isInFarcasterMiniApp(timeoutMs: number = 100): Promise<boolean> {
+  if (typeof window === "undefined") return false;
+  
+  try {
+    const sdk = await loadFarcasterSdk();
+    return await sdk.isInMiniApp();
+  } catch (error) {
+    console.warn('Farcaster SDK detection failed, using fallback:', error);
+    return fallbackDetect();
+  }
+}
+
+/**
+ * Synchronous version for backwards compatibility
+ * Note: This uses basic detection and should be replaced with async version when possible
+ */
+export function isInFarcasterMiniAppSync(): boolean {
+  if (typeof window === "undefined") return false;
+  
+  try {
+    // Basic detection without the async SDK call
+    return (
+      window.location.href.includes("farcaster") ||
+      window.location.href.includes("miniapp") ||
+      // Check if we're in an iframe which is common for miniapps
+      window.self !== window.top
+    );
+  } catch (error) {
+    return false;
+  }
+}
+
+/**
+ * Opens a URL using Farcaster's official SDK capabilities
+ * @param url - The URL to open
+ * @param fallbackToNewTab - Whether to open in new tab as fallback (default: true)
+ */
+export async function openUrlInFarcaster(url: string, fallbackToNewTab: boolean = true): Promise<void> {
+  if (typeof window === "undefined") {
+    throw new Error("URL opening is only supported in browser environments.");
+  }
```

Comment:
duplicate you already check if in farcaster app

Reply:
@vortex-hue